lib/ash/resource/validation/match.ex

```diff
@@ -55,6 +55,34 @@ defmodule Ash.Resource.Validation.Match do
     end
   end
 
+  @impl true
+  def atomic(changeset, opts, context) do
+    cond do
+      argument?(changeset.action, opts[:attribute]) ->
+        validate(changeset, opts, context)
+
+      not Ash.Changeset.changing_attribute?(changeset, opts[:attribute]) ->
```

Comment:
So, we can do the match in an expression as long as it's not a regex. What it would look like is something like this:

```elixir
argument?(...) ->
  validate(...)

!is_expr(new_value) -> # get new value from attributes[attr] || atomics[attr]
  do_match_logic_inline(new_value)
  
is_regex(opts[:match]) ->
   {:not_atomic, "can't do regexes atomically"}
 
is_expr(new_value) -> # get new value from attributes[attr] || atomics[attr]
  {:atomic, ...., expr(contains(^atomic_ref(attr), ^opts[:match]))}
```

this kind of thing. I need to write a guide on how atomics work, thats part of what I'll be doing soon.

Reply:
I'm not understanding how `:match` can be anything other than a Regex given it gets passed as second argument to `String.match?/2`, which only accepts a regex ad second argument 🤔

Another thing I've noticed is that Match returns `:ok` if the attribute [is not changing](https://github.com/ash-project/ash/blob/2a585662adbbb11c85125f01927312a69917142c/lib/ash/resource/validation/match.ex#L53), so I guess also the "not changing" case is atomic-capable

Reply:
That is wrong actually 😞it needs to be applied even if not changing.

Reply:
Should this be fixed in this PR or in a separate one?

Reply:
I think this is wrong in other validations too (e.g. OneOf does the right thing in `atomic` but only validates changes and arguments in `validate`), would it make sense to fix those in a separate PR?
This PR should be good otherwise (since it already does the right thing if the attribute is not changing, which is declaring the validation not atomic)

Reply:
Yeah, we definitely need to fix any other validations 😢 They need to check current data if things aren't changing (unless the validation is explicitly checking for changes). Also yeah you're right I don't know why I thought `String.match` accepted a string and would check for it being contained 😆 that was dumb.